wzdx/tools/cdot_geospatial_api.py

```diff
@@ -0,0 +1,272 @@
+        # https://dtdapps.coloradodot.info/arcgis/rest/services/LRS/Routes/MapServer/exts/CdotLrsAccessRounded/Route?routeId=070A&outSR=4326&f=pjson
+
+        resp = self._make_cached_web_request(url)
+        if not resp:
+            return None
+        # response = [{'routeID': '070A', 'MMin': 0, 'MMax': 499}]
+        return resp
+
+    def get_route_details(self, routeId):
+        parameters = []
+        parameters.append(f"routeId={routeId}")
+        parameters.append(f"outSR={self.SR}")
+        parameters.append(f"f=pjson")
+
+        url = f"{self.BASE_URL}/{self.GET_ROUTE_API}?{'&'.join(parameters)}"
+        logging.debug(url)
+
+        # https://dtdapps.coloradodot.info/arcgis/rest/services/LRS/Routes/MapServer/exts/CdotLrsAccessRounded/Routes?f=pjson
+        # https://dtdapps.coloradodot.info/arcgis/rest/services/LRS/Routes/MapServer/exts/CdotLrsAccessRounded/Route?routeId=070A&outSR=4326&f=pjson
+
+        resp = self._make_cached_web_request(url)
+        if not resp:
+            return None
+        # response = {'routeID': '070A', 'MMin': 0, 'MMax': 499}
+
+        route_details = {
+            "Route": resp["features"][0]["attributes"]["Route"],
+            "MMin": float(resp["features"][0]["attributes"]["MMin"]),
+            "MMax": float(resp["features"][0]["attributes"]["MMax"]),
+        }
+
+        return route_details
+
+    def get_route_and_measure(self, latLng, heading=None, tolerance=10000):
+        # Get route ID and mile marker from lat/long and heading
+        lat, lng = latLng
+
+        parameters = []
+        parameters.append(f"x={lng}")
+        parameters.append(f"y={lat}")
+        parameters.append(f"tolerance={tolerance}")
+        parameters.append(f"inSR={self.SR}")
+        parameters.append(f"outSR={self.SR}")
+        parameters.append(f"f=pjson")
+
+        url = f"{self.BASE_URL}/{self.GET_ROUTE_AND_MEASURE_API}?{'&'.join(parameters)}"
+        logging.debug(url)
+
+        # https://dtdapps.coloradodot.info/arcgis/rest/services/LRS/Routes/MapServer/exts/CdotLrsAccessRounded/MeasureAtPoint?x=-105&y=39.5&inSR=4326&tolerance=10000&outSR=&f=html
+        resp = self._make_cached_web_request(url)
+        if not resp:
+            return None
+
+        if not resp.get("features"):
+            return {}
+        route_details = {
+            "Route": resp["features"][0]["attributes"]["Route"],
+            "Measure": float(resp["features"][0]["attributes"]["Measure"]),
+            "MMin": float(resp["features"][0]["attributes"]["MMin"]),
+            "MMax": float(resp["features"][0]["attributes"]["MMax"]),
+            "Distance": float(resp["features"][0]["attributes"]["Distance"]),
+        }
+
+        if heading:
+            step = 0.1  # 500 ft
+            route = route_details["Route"]
+            measure = route_details["Measure"]
+            mMin = route_details["MMin"]
+            mMax = route_details["MMax"]
```

Comment:
nMax doesn't seem to be used, should we be checking against it?

Reply:
Added a check, and improved the logging